### PR TITLE
feat: allow gaps that are multiple of event_resolution

### PR DESCRIPTION
**Pull request by BelhsanHmida**
### Summary

This PR allows CSV uploads to contain gaps, as long as the gap duration is an integer
multiple of the intended base resolution.

### Changes

- [x] Relax frequency checks in `read_csv` to allow gaps that are multiples of the base resolution.
- [x] Add `BeliefsDataFrame.most_common_event_frequency` to infer the base resolution even
  when data contains gaps.
- [x] Add test that the frequency of a bdf with a gap is still None
- [x] Update `SensorDataFileSchema.post_load` to infer the input resolution using
  `most_common_event_frequency` instead of `event_frequency`. (handeld in flexmeasures PR [#1918](https://github.com/FlexMeasures/flexmeasures/pull/1918))

### Rationale

`event_frequency` becomes `None` for gappy data, even when the underlying resolution is
clear. This change separates regularity from intended resolution and allows valid
real-world CSV uploads without affecting existing behavior.

#### Automated tests (timely-beliefs)

Run the unit tests that cover regular and gappy data:

```bash
pytest timely_beliefs/tests/test_read_csv_frequency.py
```
These tests verify that:

- For gappy data where gaps are integer multiples of the base resolution:
    - event_frequency is considered irregular
    - most_common_event_frequency still infers the correct base resolution
- For regular data:
   - most_common_event_frequency matches event_frequency
 
## Related issue
closes [#156](https://github.com/SeitaBV/ems/issues/156)

### CODE REVIEW

File: timely_beliefs/beliefs/classes.py [hunk 1]
```diff
@@ -1219,6 +1219,28 @@ def event_frequency(self) -> timedelta | None:
...
+        if freq is not None and not pd.isna(freq):
+            return freq
+
+        event_starts = pd.to_datetime(self.event_starts.unique())
+        if len(event_starts) < 3:
+            return None
+
+        diffs = pd.Series(event_starts).sort_values().diff().dropna()
+        if diffs.empty:
+            return None
```
Comment (Flix6x) [inline timely_beliefs/beliefs/classes.py]: We cannot reach this statement, because we are guaranteed to have at least 3 unique event starts at this point, which implies that `diffs` cannot be empty.

(In such cases, rather than silently return `None`, I'd suggest raising an exception instead. However, in this particular case, I feel confident that  `diffs` cannot be empty.)

File: timely_beliefs/beliefs/classes.py [hunk 2]
```diff
@@ -1219,6 +1219,28 @@ def event_frequency(self) -> timedelta | None:
...
+    @property
+    def most_common_event_frequency(self) -> timedelta | None:
+        """Most common duration between event starts.
+
+        Unlike event_frequency, this also works when the data contains gaps,
+        as long as the gaps are integer multiples of the base resolution.
+        """
+        # If data is perfectly regular, reuse inferred frequency
+        freq = self.event_frequency
+        if freq is not None and not pd.isna(freq):
```
[the remark this answers is not in the document]
Comment (BelhsanHmida) [reply timely_beliefs/beliefs/classes.py]: yes it is.

File: timely_beliefs/beliefs/classes.py (hunk 1 above)
Comment (BelhsanHmida) [reply timely_beliefs/beliefs/classes.py]: Good catch, you’re right.
Given that `event_frequency` already handles the cases with fewer than 3 event starts, the checks for `len(event_starts) < 3` and `diffs.empty` are indeed unreachable here. I removed them to simplify the code.

File: timely_beliefs/beliefs/classes.py [hunk 3]
```diff
@@ -1219,6 +1219,28 @@ def event_frequency(self) -> timedelta | None:
...
+            return freq
+
+        event_starts = pd.to_datetime(self.event_starts.unique())
+        if len(event_starts) < 3:
+            return None
```
[the remark this answers is not in the document]
Comment (BelhsanHmida) [reply timely_beliefs/beliefs/classes.py]: good point, removed

**Comment by nhoening** [comment]
@BelhsanHmida is this ready for a final review?

**Review by Flix6x — APPROVED** [review]
Nice job!

I added a small fix (to my own code!) after a realization I had from reviewing.
